ark/model/model_graph_impl.cpp

```diff
@@ -0,0 +1,480 @@
+        }
+        for (auto &consumer : node->consumers) {
+            auto it2 = node_map.find(consumer);
+            if (it2 == node_map.end()) {
+                ERR(ModelError, "unexpected error");
+            }
+            new_node->consumers.push_back(it2->second);
+        }
+    }
+    op_to_node_.clear();
+    for (const auto &p : other.op_to_node_) {
+        auto it = node_map.find(p.second);
+        if (it == node_map.end()) {
+            ERR(ModelError, "unexpected error");
+        }
+        op_to_node_[p.first] = it->second;
+    }
+    tensor_to_producer_op_ = other.tensor_to_producer_op_;
+    rank_ = other.rank_;
+    world_size_ = other.world_size_;
+    compressed_ = other.compressed_;
+    return *this;
+}
+
+void ModelGraph::Impl::compress_nodes() {
+    if (!compressed_) {
+        this->recursive_remove_virtual_nodes();
+        this->recursive_merge_nodes();
+        compressed_ = true;
+    }
+}
+
+bool ModelGraph::Impl::verify() const {
+    for (auto &node : nodes_) {
+        if (node->ops.size() == 0) {
+            LOG(DEBUG, "node has no ops");
+            return false;
+        }
+        for (auto &op : node->ops) {
+            if (op_to_node_.find(op) == op_to_node_.end()) {
+                LOG(DEBUG, "op has not been added to the graph");
+                return false;
+            }
+            if (op_to_node_.at(op) != node) {
+                LOG(DEBUG, "op is not in the correct node");
+                return false;
+            }
+            op->verify();
+            for (auto &tns : op->result_tensors()) {
+                if (tensor_to_producer_op_.find(tns) ==
+                    tensor_to_producer_op_.end()) {
+                    LOG(DEBUG, "result tensor has not been produced by any op");
+                    return false;
+                }
+                if (tensor_to_producer_op_.at(tns) != op) {
+                    LOG(DEBUG, "result tensor has been produced by another op");
+                    return false;
+                }
+            }
+            for (auto &tns : op->input_tensors()) {
+                if (tensor_to_producer_op_.find(tns) ==
+                    tensor_to_producer_op_.end()) {
+                    LOG(DEBUG, "input tensor has not been produced by any op");
+                    return false;
+                }
+            }
+        }
+        for (auto &producer : node->producers) {
+            if (producer->consumers.find(node) == producer->consumers.end()) {
+                LOG(DEBUG, "producer does not have this node as consumer");
+                return false;
+            }
+        }
+        for (auto &consumer : node->consumers) {
+            if (consumer->producers.find(node) == consumer->producers.end()) {
+                LOG(DEBUG, "consumer does not have this node as producer");
+                return false;
+            }
+        }
+    }
+    try {
+        [[maybe_unused]] auto j = Json::parse(this->serialize(-1));
```

Comment:
Will the compiler eliminate this sentence if `j` is not used

Reply:
Modified in [ce5d87a](https://github.com/microsoft/ark/pull/205/commits/ce5d87a74087e8dfc1254a4f7c72e6be56881fab) and added a TODO.